### PR TITLE
tpm_device: automate tpm passthrough cases

### PR DESCRIPTION
**Pull request by yanqzhan**
Automate tpm passthrough cases. Need tpm.py file in
avocado-vt/pull/2011.

Signed-off-by: Yanqiu Zhang <yanqzhan@redhat.com>

### CODE REVIEW

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        else:
+            # The default tpm model is "tpm-tis"
+            pattern = "<tpm model=\"tpm-tis\">"
+        # Check tpm model
+        xml_after_adding_device = VMXML.new_from_dumpxml(vm_name)
+        if pattern not in str(xml_after_adding_device):
+            test.fail("Can not find the %s tpm device xml "
+                      "in the guest xml file." % tpm_model)
+        # Check backend type
+        pattern = "<backend type=\"%s\"" % backend_type
```
Comment (brucepatricklee) [inline libvirt/tests/src/virtual_device/tpm_device.py]: use ' instead of " could save  \

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                      "in the guest xml file." % backend_type)
+        # Check backend version
+        if backend_version:
+            pattern = "'emulator' version=\"%s\"" % backend_version
+            if pattern not in str(xml_after_adding_device):
+                test.fail("Can not find the %s backend version xml for tpm dev "
+                          "in the guest xml file." % backend_version)
+        # Check device path
+        if backend_type == "passthrough":
+            pattern = "<device path=\"/dev/tpm0\""
```
Comment (brucepatricklee) [inline libvirt/tests/src/virtual_device/tpm_device.py]: same with the previous comment

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        logging.debug("vm with tpm dev is %s", vm_xml)
+        vm_xml.sync()
+        virsh.start(vm_name, debug=True, ignore_status=False)
+        check_dumpxml()
+        check_qemu_cmd_line()
+
+    finally:
+        if vm.is_alive():
+            virsh.destroy(vm_name)
+        vm_xml_backup.sync()
```
Comment (brucepatricklee) [inline libvirt/tests/src/virtual_device/tpm_device.py]: no need to destroy vm, the sync() will take care of it

**Comment by brucepatricklee** [comment]
@yanqzhan please add the test results in the comment

**Comment by yanqzhan** [comment]
Updates:
1. Add tpm basic function tests, including tpm test in guest, and host tpm chip version identification and host tpm test before passthrough and after passthrough.
2. tpm function test differs according to different tpm version(for passthrough denpends on host tpm_real_v, for emulator depends on backend_version)
3. Some attributes(backend_version) are only for 'emulator' backend_type, which not fully tested in .py files since it's not supported temporarily so not involved in .cfg matrix. But some 'if' branches are set in .py files to filter 'passthrough' situation more accurately.

**Comment by yanqzhan** [comment]
Test results:
# avocado run --vt-type libvirt virtual_devices.tpm_device
1. A host with tpm2.0 enabled and proper pkgs version:
JOB ID     : 0e191668618f6cd4b9b8dfeb5c9b70fdbafb077f
JOB LOG    : /root/avocado/job-results/job-2019-03-27T04.27-0e19166/job.log
 (1/3) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.tpm-tis_model.passthrough.device_path_assign: PASS (28.32 s)
 (2/3) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.tpm-crb_model.passthrough.device_path_assign: PASS (27.11 s)
 (3/3) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.default_model.passthrough.device_path_default: PASS (6.95 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 64.03 s
2. A host with tpm1.2 enabled and proper pkgs version:
JOB ID     : 6b03c9adcb7980a8c4609713b02567f6f751f868
JOB LOG    : /root/avocado/job-results/job-2019-03-27T17.47-6b03c9a/job.log
 (1/3) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.tpm-tis_model.passthrough.device_path_assign: PASS (68.68 s)
 (2/3) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.tpm-crb_model.passthrough.device_path_assign: PASS (71.02 s)
 (3/3) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.default_model.passthrough.device_path_default: PASS (9.78 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 151.94 s
3. A host without  tpm chip
JOB ID     : 2003c8600d5de65e9d70d170286794aebde5ec6e
JOB LOG    : /root/avocado/job-results/job-2019-03-27T05.58-2003c86/job.log
 (1/3) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.tpm-tis_model.passthrough.device_path_assign: CANCEL: No TPM chip found on this host for passthrough testing. (6.08 s)...
4. A host with a lower qemu version which does not support tpm passthrouh
JOB ID     : ef7861d48e4ecfa6cd75e6e42674d93795087ac2
JOB LOG    : /root/avocado/job-results/job-2019-03-27T17.51-ef7861d/job.log
 (1/3) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.tpm-tis_model.passthrough.device_path_assign: CANCEL (26.63 s)...
5. A host with tpm1.2 enabled and a higher qemu which does not support tpm passthrough(rhbz#1688312)
# avocado run --vt-type libvirt  tpm_device.normal_test.tpm-tis_model.passthrough.device_path_assign
JOB ID     : 9889ab3ebf5f1087255aebefa6994dff15185f48
JOB LOG    : /root/avocado/job-results/job-2019-03-27T16.53-9889ab3/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.tpm-tis_model.passthrough.device_path_assign: ERROR: Command '/usr/bin/virsh start avocado-vt-vm1 ' failed.\nstdout: b'\n'\nstderr: b'error: Failed to start domain avocado-vt-vm1\nerror: unsupported configuration: The QEMU executable /usr/libexec/qemu-kvm does not support TPM backend type passthrough\n'\naddit... (9.77 s)
RESULTS    : PASS 0 | ERROR 1 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 11.45 s

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
+import re
+import logging
+
+from virttest.libvirt_xml.devices.tpm import Tpm
+from virttest.libvirt_xml.vm_xml import VMXML
+from virttest import virsh
+
+from provider import libvirt_version
+#from virttest import utils_misc
```
[the remark this answers is not in the document]
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Thanks. Done.

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    2. check whether the guest can be started
+    3. check the xml and qemu cmd line
+    4. check tpm usage in guest os
+    """
+    # Tpm passthrough supported since 1.0.5.
+    if not libvirt_version.version_compare(1, 0, 5):
+        test.cancel("Tpm device is not supported "
+                    "on current version.")
+
+    logging.debug("my first logging line")
```
[the remark this answers is not in the document]
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done: removed.

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 1 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 2 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 3 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    tpm_model = params.get("tpm_model")
+    backend_type = params.get("backend_type")
+    backend_version = params.get("backend_version")
+    device_path = params.get("device_path")
+
+    # Check tpm chip on host for passthrough testing
+    if backend_type == "passthrough":
+        dmesg_info = subprocess.check_output("dmesg|grep tpm -wi", shell=True).decode()
+        logging.debug("dmesg info about tpm:\n %s", dmesg_info)
+        if re.search("No TPM chip found", dmesg_info):
```
Comment (chunfuwen) [inline libvirt/tests/src/virtual_device/tpm_device.py]: 47 and 49 maybe merged into one if condition

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+                test.fail("Can not find the %s tpm device "
+                          "in qemu cmd line." % backend_type)
+            # Check device path
+            dev_num = re.search(r"\d+", device_path).group()
+            pattern = "id=tpm-tpm%s" % dev_num
+            if not re.search(pattern, cmdline):
+                test.fail("Can not find the %s tpm device "
+                          "in qemu cmd line." % device_path)
+
+    def test_host_tpm_bef(tpm_v):
```
Comment (chunfuwen) [inline libvirt/tests/src/virtual_device/tpm_device.py]: change test_host_tpm_bef to get_host_tpm_bef

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        """
+        Test host tpm function and identify its real version before passthrough
+        """
+        # Try tcsd tool for suspected tpm1.2 chip on host
+        tpm_real_v = tpm_v
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_device/tpm_device.py]: I think no need for this line.

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+                else:
+                    logging.info("Host tcsd.serivce start failed")
+            else:
+                tpm_real_v = "1.2"
+                logging.info("Host tpm version info:")
+                result = process.run("tpm_version", ignore_status=False)
+                logging.debug("[host]# tpm_version\n %s", result.stdout)
+                time.sleep(2)
+                service_mgr.stop('tcsd')
+        elif tpm_real_v != "1.2":
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_device/tpm_device.py]: And change the param to tmp_v for consistency

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 8 above)
Comment (kylazhang) [reply libvirt/tests/src/virtual_device/tpm_device.py]: And better give explanation here for why check version like this

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+            dev_num = re.search(r"\d+", device_path).group()
+            pattern = "id=tpm-tpm%s" % dev_num
+            if not re.search(pattern, cmdline):
+                test.fail("Can not find the %s tpm device "
+                          "in qemu cmd line." % device_path)
+
+    def test_host_tpm_bef(tpm_v):
+        """
+        Test host tpm function and identify its real version before passthrough
+        """
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_device/tpm_device.py]: please give parameter and return explanation

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+                test.cancel("Both tcsd and tpm2-tools can not work,"
+                            "pls check your host tpm version and test env.")
+            else:
+                tpm_real_v = "2.0"
+        return tpm_real_v
+
+    def test_host_tpm_aft(tpm_real_v):
+        """
+        Test host tpm function after passthrough
+        """
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_device/tpm_device.py]: please give parameter explanation

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+        if tpm_real_v == "1.2":
+            if service_mgr.start('tcsd'):
+                time.sleep(2)
+                service_mgr.stop('tcsd')
+                test.fail("Host tpm should not work after passthrough to guest.")
+            else:
+                logging.info("Expected failure:")
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_device/tpm_device.py]: please give the exact message what failure is expected.

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+        elif tpm_real_v == "2.0":
+            if not process.run("tpm2_getrandom 7", ignore_status=True).exit_status:
+                test.fail("Host tpm should not work after passthrough to guest.")
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_device/tpm_device.py]: no logging info for expected failure like above?

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+
+    def test_guest_tpm(expect_version, session, expect_fail):
+        """
+        Test tpm function in guest
+        """
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_device/tpm_device.py]: please give parameters explanation here

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+        session = vm.wait_for_login()
+        if expect_version == "1.2":
+            # Install tpm-tools and test by tcsd method
+            if not utils_package.package_install(["tpm-tools"], session, 360):
+                test.cancel("Failed to install tpm-tools package in guest")
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_device/tpm_device.py]: test.error

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+                        test.fail("Failed to start tcsd.service in guest")
+                else:
+                    status, output = session.cmd_status_output("tpm_version")
+                    logging.debug(output)
+                    if status:
+                        test.fail("Guest tpm can not work")
+        else:
+            # If expect_version is tpm2.0, install and test by tpm2-tools
+            if not utils_package.package_install(["tpm2-tools"], session, 360):
+                test.cancel("Failed to install tpm2-tools package in guest")
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_device/tpm_device.py]: test.error

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 6 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done:
        dmesg_error = re.search("No TPM chip found|TPM is disabled", dmesg_info)
        if dmesg_error:
            test.cancel(dmesg_error.group())

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 7 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+import subprocess
+import time
+
+from virttest.libvirt_xml.devices.tpm import Tpm
+from virttest.libvirt_xml.vm_xml import VMXML
+from virttest import virsh
+from virttest import utils_package
+
+from provider import libvirt_version
+from virttest import utils_misc
```
[the remark this answers is not in the document]
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+                    "on current libvirt version.")
+    # Tpm passthrough supported since qemu 2.12.0-49.
+    if not utils_misc.compare_qemu_version(2, 9, 0, is_rhev=False):
+        test.cancel("Tpm device is not supported "
+                    "on current qemu version.")
+
+    vm_name = params.get("main_vm", "avocado-vt-vm1")
+    vm = env.get_vm(vm_name)
+
+    status_error = params.get("status_error", "no") == "yes"
```
[the remark this answers is not in the document]
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+    tpm_model = params.get("tpm_model")
+    backend_type = params.get("backend_type")
+    backend_version = params.get("backend_version")
+    device_path = params.get("device_path")
+
+    # Check tpm chip on host for passthrough testing
+    if backend_type == "passthrough":
+        dmesg_info = subprocess.check_output("dmesg|grep tpm -wi", shell=True).decode()
```
[the remark this answers is not in the document]
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+        elif re.search("TPM is disabled", dmesg_info):
+            test.cancel("TPM is disabled, pls enable in BIOS for passthrough testing.")
+        else:
+            # Try to check host tpm chip version
+            tpm_v = None
+            if re.search("2.0 TPM", dmesg_info):
+                tpm_v = "2.0"
+                if not utils_package.package_install("tpm2-tools"):
+                    # package_install() return 'True' if succeed
+                    test.cancel("Failed to install tpm2-tools on host")
```
[the remark this answers is not in the document]
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+            else:
+                if re.search("1.2 TPM", dmesg_info):
+                    tpm_v = "1.2"
+                # If "1.2 TPM" or no version info in dmesg, try to test a tpm1.2 at first
+                if not utils_package.package_install("tpm-tools"):
+                    test.cancel("Failed to install tpm-tools on host")
```
[the remark this answers is not in the document]
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 12 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 13 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 14 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done:
        :param expect_version: guest tpm version, as host version, or emulator specified
        :param session: Guest session to be tested
        :param expect_fail: guest tpm is expectedly fail to work

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 15 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 16 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

**Comment by yanqzhan** [comment]
Add 2 new negative scenarios, need vm_xml.py modified by avocado-vt/pull/2044.
And all local run passed.
# avocado run --vt-type libvirt virtual_devices.tpm_device.negative_test.reuse_passthrgh_tpm
JOB ID     : 1bca36b99d95e84cd5b957ade5a86199b177d8a8
JOB LOG    : /root/avocado/job-results/job-2019-04-19T08.47-1bca36b/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.negative_test.reuse_passthrgh_tpm: PASS (27.71 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 29.36 s
(.libvirt-ci-venv-ci-runtest-SmXw37) [root@dell-per740-07 ~]# vim /root/avocado/job-results/job-2019-04-19T08.47-1bca36b/job.log
# avocado list --vt-type libvirt virtual_devices.tpm_device
VT type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.tpm-tis_model.passthrough.device_path_assign
VT type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.tpm-crb_model.passthrough.device_path_assign
VT type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.default_model.passthrough.device_path_default
VT type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.negative_test.reuse_passthrgh_tpm
VT type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.negative_test.multi_passthrgh_tpm
# avocado run --vt-type libvirt virtual_devices.tpm_device.negative_test.multi_passthrgh_tpm
JOB ID     : 0b1ad6582d62b8aacde5617c6ac228f99f0fc59e
JOB LOG    : /root/avocado/job-results/job-2019-04-19T08.50-0b1ad65/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.negative_test.multi_passthrgh_tpm: PASS (7.44 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 9.09 s
(.libvirt-ci-venv-ci-runtest-SmXw37) [root@dell-per740-07 ~]# vim /root/avocado/job-results/job-2019-04-19T08.50-0b1ad65/job.log
(.libvirt-ci-venv-ci-runtest-SmXw37) [root@dell-per740-07 ~]# avocado run --vt-type libvirt virtual_devices.tpm_device.normal_test
JOB ID     : edfe20c3dc0b38492e5da2eecc8de0d607208dad
JOB LOG    : /root/avocado/job-results/job-2019-04-19T08.58-edfe20c/job.log
 (1/3) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.tpm-tis_model.passthrough.device_path_assign: PASS (26.51 s)
 (2/3) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.tpm-crb_model.passthrough.device_path_assign: PASS (27.62 s)
 (3/3) type_specific.io-github-autotest-libvirt.virtual_devices.tpm_device.normal_test.default_model.passthrough.device_path_default: PASS (7.29 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 63.17 s

**Comment by yanqzhan** [comment]
@kylazhang @chunfuwen @brucepatricklee 
Thank you for your review. Code modified, and new scenarios added. Pls help review again, including avocado-vt/pull/2011, avocado-vt/pull/2044. Thank you!

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 22]
```diff
@@ -0,0 +1,321 @@
...
+        vm2_xml.add_device(tpm_dev)
+        vm2_xml.sync()
+        ret = virsh.start(vm2_name, ignore_status=True, debug=True)
+        if ret:
+            logging.info("Expected failure when try to passthrough a tpm"
+                         " that being used by another guest")
+            return
+        else:
+            test.fail("Reuse a passthroughed tpm should not succeed.")
+        logging.info("------DONE for tpm reuse try------")
```
Comment (brucepatricklee) [inline libvirt/tests/src/virtual_device/tpm_device.py]: this line will never got change to run, am I right?

Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: thx, deleted.

File: libvirt/tests/src/virtual_device/tpm_device.py [hunk 23]
```diff
@@ -0,0 +1,321 @@
...
+    def test_guest_tpm(expect_version, session, expect_fail):
+        """
+        Test tpm function in guest
+
+        :param expect_version: guest tpm version, as host version, or emulator specified
+        :param session: Guest session to be tested
+        :param expect_fail: guest tpm is expectedly fail to work
+        """
+        logging.info("------Checking guest tpm device work------")
+        session = vm.wait_for_login()
```
[the remark this answers is not in the document]
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: deleted.

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (kylazhang) [reply libvirt/tests/src/virtual_device/tpm_device.py]: still not changed here

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 10 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

File: libvirt/tests/src/virtual_device/tpm_device.py (hunk 11 above)
Comment (yanqzhan) [reply libvirt/tests/src/virtual_device/tpm_device.py]: Done.

**Pull request merged**